src/pages/api/auth/[...nextauth].ts

```diff
@@ -0,0 +1,115 @@
+import NextAuth from 'next-auth';
+import { FirestoreAdapter } from '@auth/firebase-adapter';
+import GoogleProvider from 'next-auth/providers/google';
+import CredentialsProvider from 'next-auth/providers/credentials';
+import firebase, { auth } from '@/firebase/client';
+import { cert } from 'firebase-admin/app';
+
+import {
+  createUserWithEmailAndPassword,
+  signInWithEmailAndPassword,
+} from 'firebase/auth';
+
+const GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID;
+const GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET;
+
+if (!GOOGLE_CLIENT_ID || !GOOGLE_CLIENT_SECRET) {
+  throw new Error('GOOGLE_CLIENT_ID or GOOGLE_CLIENT_SECRET not set');
+}
+
+const adapter = FirestoreAdapter({
+  credential: cert({
+    projectId: process.env.FIREBASE_PROJECT_ID,
+    clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
+    privateKey: process.env.FIREBASE_PRIVATE_KEY
+      ? JSON.parse(process.env.FIREBASE_PRIVATE_KEY)
+      : undefined,
+  }),
+});
+export default NextAuth({
+  secret: process.env.AUTH_SECRET,
+  adapter,
+  providers: [
+    CredentialsProvider({
+      name: 'Credentials',
+      credentials: {
+        firstname: {
+          label: 'First name',
+          type: 'text',
+        },
+        lastname: {
+          label: 'Last name',
+          type: 'text',
+        },
+        email: {
+          label: 'Email',
+          type: 'email',
+          placeholder: 'jsmith@example.com',
+        },
+        password: {
+          label: 'Password',
+          type: 'password',
+        },
+      },
+      // @ts-expect-error AdaptUser returned is ok
+      async authorize(credentials) {
+        const { firstname, lastname, password, email } = credentials as Record<
+          string,
+          string
+        >;
+
+        if (firstname && lastname && password && email) {
+          // Registration
+
+          try {
+            const data = await createUserWithEmailAndPassword(
+              auth,
+              credentials?.email as string,
+              credentials?.password as string
+            );
+            // if everything is good, create and return user
+            await adapter.createUser?.({
+              name: `${firstname} ${lastname}`,
+              email,
+              emailVerified: data.user.emailVerified ? new Date() : null,
+            });
+          } catch (err) {
+            throw new Error((err as firebase.FirebaseError).code);
+          }
+
+          return (await adapter.getUserByEmail?.(email)) ?? null;
```

Comment:
We should cut registration logic out from NextAuth, because this package is managing sessions.
I mean: signin, jwt tokens, signout and that's it. Registration is deprecated from NextAuth on purpose.

Reply:
I wanted to keep this in one place, but sounds good. I will create a separate api endpoint.